luarules/gadgets/cmd_factory_stop_production.lua

```diff
@@ -41,86 +27,121 @@ for udid = 1, #UnitDefs do
+			-- 	count = count - 20
+			-- elseif count >= 5 then
+			-- 	opts = { "shift" }
+			-- 	count = count - 5
+			-- else
+			-- 	count = count - 1
+			-- end
+
+			spGiveOrderToUnit(unitID, -buildDefID, EMPTY, DEQUEUE_OPTS)
+		end
 	end
 
-	-- Dequeue build order by sending build command to factory to minimize number of commands sent
-	-- As opposed to removing each build command individually
-	local queue = spGetRealBuildQueue(unitID)
+	function gadget:AllowCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams, cmdOptions)
+		if (cmdID ~= CMD_STOP_PRODUCTION) or not isFactory[unitDefID] then
+			return true
+		end
+
+		-- Dequeue build order by sending build command to factory to minimize number of commands sent
+		-- As opposed to removing each build command individually
+		local queue = spGetRealBuildQueue(unitID)
 
-	if queue ~= nil then
-		for _, buildPair in ipairs(queue) do
-			local buildUnitDefID, count = next(buildPair, nil)
+		if queue ~= nil then
+			for _, buildPair in ipairs(queue) do
+				local buildUnitDefID, count = next(buildPair, nil)
 
-			orderDequeue(unitID, buildUnitDefID, count)
+				orderDequeue(unitID, buildUnitDefID, count)
+			end
 		end
+
+		spGiveOrderToUnit(unitID, CMD_WAIT, EMPTY, 0) -- Removes wait if there is a wait but doesn't readd it.
+		spGiveOrderToUnit(unitID, CMD_WAIT, EMPTY, 0) -- If a factory is waiting, it will not clear the current build command, even if the cmd is removed.
+		-- See: http://zero-k.info/Forum/Post/237176#237176 for details.
+		sendToUnsynced(magicValue, unitID, unitDefID, unitTeam, cmdID)
 	end
 
-	spGiveOrderToUnit(unitID, CMD_WAIT, EMPTY, 0) -- Removes wait if there is a wait but doesn't readd it.
-	spGiveOrderToUnit(unitID, CMD_WAIT, EMPTY, 0) -- If a factory is waiting, it will not clear the current build command, even if the cmd is removed.
-	-- See: http://zero-k.info/Forum/Post/237176#237176 for details.
-	return true
-end
+	--------------------------------------------------------------------------------
+	--------------------------------------------------------------------------------
+	-- Add the command to factories
 
---------------------------------------------------------------------------------
---------------------------------------------------------------------------------
--- Add the command to factories
+	function gadget:UnitCreated(unitID, unitDefID)
+		if isFactory[unitDefID] then
+			spInsertUnitCmdDesc(unitID, stopProductionCmdDesc)
+		end
+	end
 
-function gadget:UnitCreated(unitID, unitDefID)
-	if isFactory[unitDefID] then
-		spInsertUnitCmdDesc(unitID, stopProductionCmdDesc)
+	function gadget:Initialize()
+		gadgetHandler:RegisterCMDID(CMD_STOP_PRODUCTION)
+		for _, unitID in pairs(Spring.GetAllUnits()) do
+			gadget:UnitCreated(unitID, Spring.GetUnitDefID(unitID))
+		end
+	end
+else
+	local scriptUnitCommand = Script.LuaUI.UnitCommand
+
+	local spGetSpectatingState = Spring.GetSpectatingState
+	local spAreTeamsAllied = Spring.AreTeamsAllied
```

Comment:
Do these actually need to be localized? Are they being called multiple times per frame?